components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAFederatedAuthenticator.java

```diff
@@ -212,31 +212,46 @@ private IWAAuthenticatedUserBean userInformationInListedUserStores(String authen
                                                                        String userStoreDomains)
             throws AuthenticationFailedException {
 
-        String userStoreDomainForUser = null;
         IWAAuthenticatedUserBean authenticatedUserBean = new IWAAuthenticatedUserBean();
-        try {
-            for (String userStoreDomain : userStoreDomains.split(",")) {
-                resolveUserFromUserStore(
-                        authenticatedUserBean, authenticatedUserName, tenantDomain, userStoreDomain.trim());
-                if (authenticatedUserBean.isUserExists()) {
-                    userStoreDomainForUser = userStoreDomain.trim();
-                    break;
+        authenticatedUserBean.setTenantDomain(tenantDomain);
+        authenticatedUserBean.setUser(authenticatedUserName);
+        MultiAttributeLoginService multiAttributeLoginService =
+                IWAServiceDataHolder.getInstance().getMultiAttributeLoginService();
+
+        if (multiAttributeLoginService.isEnabled(tenantDomain)) {
+            ResolvedUserResult resolvedUserResult = multiAttributeLoginService.resolveUser(
+                    authenticatedUserName, tenantDomain);
+            if (resolvedUserResult != null &&
+                    ResolvedUserResult.UserResolvedStatus.SUCCESS.equals(resolvedUserResult.getResolvedStatus()) &&
+                    userStoreDomains.toUpperCase().contains(resolvedUserResult.getUser().getUserStoreDomain())) {
+                authenticatedUserBean.setUserExists(true);
+                authenticatedUserBean.setUserStoreDomain(resolvedUserResult.getUser().getUserStoreDomain());
+                if (StringUtils.isNotBlank(resolvedUserResult.getUser().getPreferredUsername())) {
+                    authenticatedUserBean.setUser(resolvedUserResult.getUser().getPreferredUsername());
                 }
             }
+        } else {
+            boolean isUserExists = false;
+            String userStoreDomainForUser = null;
+            try {
+                for (String userStoreDomain : userStoreDomains.split(",")) {
+                    if (isUserExistsInUserStore(authenticatedUserName, tenantDomain, userStoreDomain.trim())) {
+                        isUserExists = true;
+                        userStoreDomainForUser = userStoreDomain.trim();
+                        break;
+                    }
+                }
+                authenticatedUserBean.setUserExists(isUserExists);
+                authenticatedUserBean.setUserStoreDomain(userStoreDomainForUser);
 
-            authenticatedUserBean.setTenantDomain(tenantDomain);
-            if (StringUtils.isEmpty(authenticatedUserBean.getUser())) {
-                authenticatedUserBean.setUser(authenticatedUserName);
+            } catch (AuthenticationFailedException e) {
+                String msg = "IWAApplicationAuthenticator failed to find the user:%s of tenantDomain=%s in neither " +
```

Comment:
Don't we have error constant class to maintain these error messages and proper codes ?

Reply:
Error messages are not maintained in a constant class in this component

Reply:
Shall we add an error class with descriptions and codes ?